File: web/src/store/entities.ts
import { Exocore, exocore, fromProtoTimestamp, MutationBuilder } from 'exocore';
import { exomind } from '../protos';

export class EntityTraits {
    public entity: exocore.store.IEntity;

    private typeTraits: { [type: string]: exocore.store.ITrait[] } = {}
    private idTraits: { [id: string]: exocore.store.ITrait } = {}
    private idInstances: { [id: string]: EntityTrait<unknown> } = {}
    private priorityTraitId?: string

    constructor(entity: exocore.store.IEntity) {
        this.entity = entity;

        // check if it's a special entity (ex: inbox)
        let priorityTrait: [exocore.store.ITrait, ITraitConstants] = null;
        if (this.entity.id in TRAITS_CONSTANTS) {
            const traitsConsts = TRAITS_CONSTANTS[this.entity.id];
            for (const trait of this.entity.traits) {
                if (trait.id == this.entity.id) {
                    priorityTrait = [trait, traitsConsts];
                    break;
                }
            }
        }

        // index traits by ids and types
        for (const trait of this.entity.traits) {
            const msgType = Exocore.registry.canonicalFullName(trait.message.type_url);
            if (!(msgType in this.typeTraits)) {
                this.typeTraits[msgType] = [trait];
            } else {
                this.typeTraits[msgType].push(trait);
            }

            this.idTraits[trait.id] = trait;

            let traitConsts;
            if (this.entity.id == trait.id && this.entity.id in TRAITS_CONSTANTS) {
                // special entity
                traitConsts = TRAITS_CONSTANTS[this.entity.id];
            } else if (msgType in TRAITS_CONSTANTS) {
                traitConsts = TRAITS_CONSTANTS[msgType];
            }

            if (traitConsts && ((priorityTrait == null || traitConsts.order < priorityTrait[1].order))) {
                priorityTrait = [trait, traitConsts];
            }
        }

        if (priorityTrait != null) {
            this.priorityTraitId = priorityTrait[0].id;
        }
    }

    get id(): string {
        return this.entity.id;
    }

    traitOfType<T>(msg: unknown): EntityTrait<T> | null {
        const traits = this.traitsOfType<T>(msg);
        if (traits.length > 0) {
            return traits[0];
        } else {
            return null;
        }
    }

    traitsOfType<T>(msg: unknown): EntityTrait<T>[] {
        const msgType = Exocore.registry.messageFullName(msg);
        return this.entity.traits.flatMap((trait: exocore.store.ITrait) => {
            if (trait.message.type_url.endsWith(msgType)) {
                return [this.trait(trait.id)];
            } else {
                return [];
            }
        });
    }

    trait<T>(id: string): EntityTrait<T> | null {
        const trait = this.idTraits[id];
        if (!trait) {
            return null;
        }

        if (!(id in this.idInstances)) {
            const et = new EntityTrait<unknown>(
                this,
                trait,
                Exocore.registry.unpackAny(trait.message),
            );
            this.idInstances[id] = et;

            return et as EntityTrait<T>;
        }

        return this.idInstances[id] as EntityTrait<T>;
    }

    get priorityTrait(): EntityTrait<unknown> | null {
        if (this.priorityTraitId != null) {
            return this.trait(this.priorityTraitId);
        } else {
            return null;
        }
    }

    priorityMatch(matcher: ITraitMatcher): unknown {
        const priorityTrait = this.priorityTrait;
        if (priorityTrait == null) {
            return null;
        }

        return priorityTrait.match(matcher);
    }
}

export class EntityTrait<T> {
    trait: exocore.store.ITrait;
    message: T;
    et: EntityTraits;

    constructor(et: EntityTraits, trait: exocore.store.ITrait, message: T) {
        this.et = et;
        this.trait = trait;
        this.message = message;
    }

    get id(): string {
        return this.trait.id;
    }

    get constants(): ITraitConstants {
        const msgType = Exocore.registry.messageFullName(this.message);
        if (this.trait.id == this.et.id && this.et.id in TRAITS_CONSTANTS) {
            // special entity
            return TRAITS_CONSTANTS[this.et.id];
        } else if (msgType in TRAITS_CONSTANTS) {
            return TRAITS_CONSTANTS[msgType];
        } else {
            return TRAITS_CONSTANTS['unknown'];
        }
    }

    get creationDate(): Date | null {
        if (this.trait.creationDate != null) {
            return fromProtoTimestamp(this.trait.creationDate);
        } else {
            return null;
        }
    }

    get modificationDate(): Date | null {
        if (this.trait.modificationDate != null) {
            return fromProtoTimestamp(this.trait.modificationDate);
        } else {
            return null;
        }
    }

    get displayName(): string {
        if (this.constants.name) {
            return this.constants.name;
        }

        if (this.constants.name_field) {
            const dict = this.message as unknown as { [p: string]: string; };
            const name = dict[this.constants.name_field];
            return name ?? this.constants.name_default ?? '*UNTITLED*';
        }

        return '*UNTITLED*';
    }

    get icon(): string {
        return this.constants.icon;
    }

    match(matcher: ITraitMatcher): unknown {
        if (this.constants.key == 'exomind.base.EmailThread' && matcher.emailThread) {
            return matcher.emailThread(this);
        } else if (this.constants.key == 'exomind.base.Email' && matcher.email) {
            return matcher.email(this);
        } else if (this.constants.key == 'exomind.base.DraftEmail' && matcher.draftEmail) {
            return matcher.draftEmail(this);
        } else if (this.constants.key == 'exomind.base.Collection' || (this.constants?.collectionLike ?? false) && matcher.collection) {
            return matcher.collection(this);
        } else if (this.constants.key == 'exomind.base.Task' && matcher.task) {
            return matcher.task(this);
        } else if (this.constants.key == 'exomind.base.Note' && matcher.note) {
            return matcher.note(this);
        } else if (this.constants.key == 'exomind.base.Link' && matcher.link) {
            return matcher.link(this);
        } else if (matcher.default) {
            return matcher.default();
        }
    }

    get canRename(): boolean {
        return !!this.constants.rename;
    }

    async rename(newName: string): Promise<exocore.store.IMutationResult> {
        if (!this.constants.rename) {
            return;
        }

        this.constants.rename(this.message, newName);
        await Exocore.store.mutate(
            MutationBuilder
                .updateEntity(this.et.id)
                .putTrait(this.message, this.id)
                .build()
        );
    }
}

export interface ITraitMatcher {
    emailThread?: (trait: EntityTrait<exomind.base.IEmailThread>) => unknown;
    email?: (trait: EntityTrait<exomind.base.IEmail>) => unknown;
    draftEmail?: (trait: EntityTrait<exomind.base.IDraftEmail>) => unknown;
    collection?: (trait: EntityTrait<exomind.base.ICollection>) => unknown;
    task?: (trait: EntityTrait<exomind.base.ITask>) => unknown;
    note?: (trait: EntityTrait<exomind.base.INote>) => unknown;
    link?: (trait: EntityTrait<exomind.base.ILink>) => unknown;
    default?: () => unknown;
}

export interface ITraitConstants {
    key: string;
    name?: string;
    name_field?: string;
    name_default?: string;
    icon: string;
    color: number;
    order: number;
    collectionLike?: boolean;
    rename?: (trait: unknown, newName: string) => void;
}

export const TRAITS_CONSTANTS: { [type: string]: ITraitConstants } = {
    'inbox': {
        key: 'inbox',
        name: 'Inbox',
        icon: 'inbox',
        collectionLike: true,
        color: 4,
        order: 0
    },
    'favorites': {
        key: 'favorites',
        name: 'Favorites',
        icon: 'star',
        collectionLike: true,
        color: 4,
        order: 1
    },
    'exomind.base.EmailThread': {
        key: 'exomind.base.EmailThread',
        name_field: 'subject',
        name_default: 'Untitled email',
        icon: 'envelope-o',
        color: 1,
        order: 2
    },
    'exomind.base.DraftEmail': {
        key: 'exomind.base.DraftEmail',
        name_field: 'subject',
        name_default: 'Untitled email',
        icon: 'envelope-o',
        color: 6,
        order: 3
    },
    'exomind.base.Email': {
        key: 'exomind.base.Email',
        name_field: 'subject',
        name_default: 'Untitled email',
        icon: 'envelope-o',
        color: 6,
        order: 4
    },
    'exomind.base.Collection': {
        key: 'exomind.base.Collection',
        name_field: 'name',
        name_default: 'Untitled collection',
        icon: 'folder-o',
        color: 2,
        order: 5,
        rename: (entity: unknown, newName: string): void => {
            const collection = entity as exomind.base.ICollection;
            collection.name = newName;
        },
    },
    'exomind.base.Task': {
        key: 'exomind.base.Task',
        name_field: 'title',
        name_default: 'Untitled task',
        icon: 'check-square-o',
        color: 7,
        order: 6,
        rename: (entity: unknown, newName: string): void => {
            const task = entity as exomind.base.ITask;
            task.title = newName;
        },
    },
    'exomind.base.Note': {
        key: 'exomind.base.Note',
        name_field: 'title',
        name_default: 'Untitled note',
        icon: 'pencil',
        color: 3,
        order: 7,
        rename: (entity: unknown, newName: string): void => {
            const note = entity as exomind.base.INote;
            note.title = newName;
        },
    },
    'exomind.base.Link': {
        key: 'exomind.base.Link',
        name_field: 'title',
        name_default: 'Untitled link',
        icon: 'link',
        color: 9,
        order: 8
    },
    'unknown': {
        key: 'unknown',
        name_field: '*UNKNOWN*',
        icon: 'question',
        color: 0,
        order: 9
    }
};

File: ios/Exomind/Entities.swift
import Foundation
import Exocore
import SwiftProtobuf

typealias EntityId = String
typealias TraitId = String

extension Exocore_Index_Entity {
    func toExtension() -> EntityExt {
        EntityExt(entity: self)
    }
}

class EntityExt {
    let inner: Exocore_Index_Entity;

    private let idTraits: [TraitId: Exocore_Index_Trait]
    private let typeTraits: [TraitId: [Exocore_Index_Trait]]
    private var traitInstances: [TraitId: Message] = [:]
    private let _priorityTrait: Exocore_Index_Trait?;

    var creationDate: Date
    var modificationDate: Date?
    var anyDate: Date?

    init(entity: Exocore_Index_Entity) {
        self.inner = entity

        var priorityTrait: (Exocore_Index_Trait, TraitConstants)?

        // check if it's a special entity (ex: inbox)
        if let entityConstants = TraitsConstants[entity.id] {
            for trait in entity.traits {
                if trait.id == entity.id {
                    priorityTrait = (trait, entityConstants)
                    break
                }
            }
        }

        var oldestDate: Date?
        var newestDate: Date?

        // index traits by ids and types
        var idTraits: [String: Exocore_Index_Trait] = [:]
        var typeTraits: [String: [Exocore_Index_Trait]] = [:]
        for trait in entity.traits {
            idTraits[trait.id] = trait

            let traitType = trait.canonicalFullName()
            var traits = typeTraits[traitType] ?? []
            traits.append(trait)
            typeTraits[traitType] = traits

            let traitConstants: TraitConstants?;
            if let entityConsts = TraitsConstants[entity.id], entity.id == trait.id {
                traitConstants = entityConsts
            } else {
                traitConstants = TraitsConstants[traitType]
            }

            if let traitConstants = traitConstants, (priorityTrait == nil || traitConstants.order < priorityTrait!.1.order) {
                priorityTrait = (trait, traitConstants)
            }

            // TODO: Should be moved to exocore
            let creationDate = trait.creationDate.date
            let modificationDate = trait.modificationDate.date
            if trait.hasCreationDate && creationDate < oldestDate ?? Date() {
                oldestDate = creationDate
            }
            if trait.hasModificationDate && modificationDate > newestDate ?? Date.init(milliseconds: 0) {
                newestDate = modificationDate
            }
        }
        self.idTraits = idTraits
        self.typeTraits = typeTraits
        self._priorityTrait = priorityTrait.map({ $0.0 })
        self.creationDate = oldestDate ?? Date()
        self.modificationDate = newestDate
        self.anyDate = self.modificationDate ?? self.creationDate
    }

    var id: EntityId {
        get {
            self.inner.id
        }
    }

    func trait<T: Message>(withId id: TraitId) -> TraitInstance<T>? {
        guard let trait = self.idTraits[id] else {
            return nil
        }

        if let message = self.traitInstances[id] as? T {
            return TraitInstance(entity: self, trait: trait, message: message)
        }

        guard let message = try? T.init(unpackingAny: trait.message) else {
            return nil
        }
        self.traitInstances[id] = message

        return TraitInstance(entity: self, trait: trait, message: message)
    }

    func trait(anyWithId id: TraitId) -> AnyTraitInstance? {
        guard let trait = self.idTraits[id],
              let traitConstants = TraitsConstants[trait.canonicalFullName()] else {
            return nil
        }

        switch traitConstants.traitType {
        case .inbox:
            let trait: TraitInstance<Exomind_Base_Collection>? = self.trait(withId: trait.id)
            return trait
        case .favorites:
            let trait: TraitInstance<Exomind_Base_Collection>? = self.trait(withId: trait.id)
            return trait
        case .emailThread:
            let trait: TraitInstance<Exomind_Base_EmailThread>? = self.trait(withId: trait.id)
            return trait
        case .draftEmail:
            let trait: TraitInstance<Exomind_Base_DraftEmail>? = self.trait(withId: trait.id)
            return trait
        case .email:
            let trait: TraitInstance<Exomind_Base_Email>? = self.trait(withId: trait.id)
            return trait
        case .collection:
            let trait: TraitInstance<Exomind_Base_Collection>? = self.trait(withId: trait.id)
            return trait
        case .task:
            let trait: TraitInstance<Exomind_Base_Task>? = self.trait(withId: trait.id)
            return trait
        case .note:
            let trait: TraitInstance<Exomind_Base_Note>? = self.trait(withId: trait.id)
            return trait
        case .link:
            let trait: TraitInstance<Exomind_Base_Link>? = self.trait(withId: trait.id)
            return trait
        }
    }

    func traitOfType<T: Message>(_ message: T.Type) -> TraitInstance<T>? {
        let traits = self.typeTraits[message.protoMessageName] ?? []
        return traits.compactMap({ (trait) -> TraitInstance<T>? in
            self.trait(withId: trait.id)
        }).first
    }

    func traitsOfType<T: Message>(_ message: T.Type) -> [TraitInstance<T>] {
        let traits = self.typeTraits[message.protoMessageName] ?? []
        return traits.compactMap({ (trait) -> TraitInstance<T>? in
            self.trait(withId: trait.id)
        })
    }

    lazy var priorityTrait: AnyTraitInstance? = {
        guard let trait = self._priorityTrait else {
            return nil
        }

        return self.trait(anyWithId: trait.id)
    }()
}

extension EntityExt: Equatable {
    static func ==(lhs: EntityExt, rhs: EntityExt) -> Bool {
        lhs.id == rhs.id && lhs.anyDate == rhs.anyDate
    }
}

protocol AnyTraitInstance {
    var entity: EntityExt? { get }
    var id: TraitId { get }
    var trait: Exocore_Index_Trait { get }
    var constants: TraitConstants? { get }
    var type: TraitType? { get }
    var displayName: String { get }
    var creationDate: Date { get }
    var modificationDate: Date? { get }

    func typeInstance() -> TraitTypeInstance?
}

struct TraitInstance<T: Message>: AnyTraitInstance {
    weak var entity: EntityExt?;
    let trait: Exocore_Index_Trait
    let message: T
    let displayName: String
    let constants: TraitConstants?

    init(entity: EntityExt, trait: Exocore_Index_Trait, message: T) {
        self.entity = entity
        self.trait = trait
        self.message = message
        self.constants = getTraitConstants(entity: entity, trait: trait)

        if let constants = self.constants {
            self.displayName = TraitInstance.getDisplayName(constants: constants, message: message)
        } else {
            self.displayName = "*UNKNOWN*"
        }
    }

    var id: TraitId {
        get {
            self.trait.id
        }
    }

    var creationDate: Date {
        get {
            self.trait.creationDate.date
        }
    }

    var modificationDate: Date? {
        get {
            if self.trait.hasModificationDate {
                return self.trait.modificationDate.date
            }

            return nil
        }
    }

    var anyDate: Date {
        get {
            self.modificationDate ?? self.creationDate
        }
    }

    var type: TraitType? {
        get {
            self.constants?.traitType
        }
    }

    func typeInstance() -> TraitTypeInstance? {
        guard let constants = self.constants else {
            return nil
        }

        switch constants.traitType {
        case .inbox:
            return .inbox(trait: self as! TraitInstance<Exomind_Base_Collection>)
        case .favorites:
            return .favorites(trait: self as! TraitInstance<Exomind_Base_Collection>)
        case .emailThread:
            return .emailThread(trait: self as! TraitInstance<Exomind_Base_EmailThread>)
        case .draftEmail:
            return .draftEmail(trait: self as! TraitInstance<Exomind_Base_DraftEmail>)
        case .email:
            return .email(trait: self as! TraitInstance<Exomind_Base_Email>)
        case .collection:
            return .collection(trait: self as! TraitInstance<Exomind_Base_Collection>)
        case .task:
            return .task(trait: self as! TraitInstance<Exomind_Base_Task>)
        case .note:
            return .note(trait: self as! TraitInstance<Exomind_Base_Note>)
        case .link:
            return .link(trait: self as! TraitInstance<Exomind_Base_Link>)
        }
    }

    func toAny() -> AnyTraitInstance? {
        self.entity?.trait(anyWithId: self.id)
    }

    private static func getDisplayName<M: Message>(constants: TraitConstants, message: M) -> String {
        if let name = constants.name {
            return name
        }

        var name: String?;
        switch message {
        case let emailThread as Exomind_Base_EmailThread:
            name = notEmpty(emailThread.subject) ?? "Untitled email"
        case let draftEmail as Exomind_Base_DraftEmail:
            name = notEmpty(draftEmail.subject) ?? "Untitled email"
        case let email as Exomind_Base_Email:
            name = notEmpty(email.subject) ?? "Untitled email"
        case let collection as Exomind_Base_Collection:
            name = collection.name
        case let task as Exomind_Base_Task:
            name = notEmpty(task.title) ?? "Untitled task"
        case let note as Exomind_Base_Note:
            name = notEmpty(note.title) ?? "Untitled note"
        case let link as Exomind_Base_Link:
            name = notEmpty(link.title) ?? "Untitled link"
        default:
            name = nil
        }

        return name ?? constants.nameDefault ?? "*UNKNOWN*"
    }
}

func notEmpty(_ str: String) -> String? {
    if str == "" {
        return nil
    }

    return str
}

enum TraitType {
    case inbox
    case favorites
    case emailThread
    case draftEmail
    case email
    case collection
    case task
    case note
    case link
}

enum TraitTypeInstance {
    case inbox(trait: TraitInstance<Exomind_Base_Collection>)
    case favorites(trait: TraitInstance<Exomind_Base_Collection>)
    case emailThread(trait: TraitInstance<Exomind_Base_EmailThread>)
    case draftEmail(trait: TraitInstance<Exomind_Base_DraftEmail>)
    case email(trait: TraitInstance<Exomind_Base_Email>)
    case collection(trait: TraitInstance<Exomind_Base_Collection>)
    case task(trait: TraitInstance<Exomind_Base_Task>)
    case note(trait: TraitInstance<Exomind_Base_Note>)
    case link(trait: TraitInstance<Exomind_Base_Link>)
}

struct TraitConstants {
    let key: String
    let traitType: TraitType
    let name: String?
    let nameDefault: String?
    let icon: String
    let color: Int
    let order: Int
    let collectionLike: Bool
}

func getTraitConstants(entity: EntityExt, trait: Exocore_Index_Trait) -> TraitConstants? {
    if let entityConsts = TraitsConstants[entity.id], entity.id == trait.id {
        return entityConsts
    } else {
        return TraitsConstants[trait.canonicalFullName()]
    }
}

let TraitsConstants: [String: TraitConstants] = [
    "inbox": TraitConstants(
            key: "inbox",
            traitType: .inbox,
            name: "Inbox",
            nameDefault: nil,
            icon: "inbox",
            color: 4,
            order: 0,
            collectionLike: true
    ),
    "favorites": TraitConstants(
            key: "favorites",
            traitType: .favorites,
            name: "Favorites",
            nameDefault: nil,
            icon: "star",
            color: 4,
            order: 1,
            collectionLike: true
    ),
    "exomind.base.EmailThread": TraitConstants(
            key: "exomind.base.EmailThread",
            traitType: .emailThread,
            name: nil,
            nameDefault: "Untitled email",
            icon: "envelope-o",
            color: 1,
            order: 2,
            collectionLike: false
    ),
    "exomind.base.DraftEmail": TraitConstants(
            key: "exomind.base.DraftEmail",
            traitType: .draftEmail,
            name: nil,
            nameDefault: "Untitled email",
            icon: "envelope-o",
            color: 6,
            order: 3,
            collectionLike: false
    ),
    "exomind.base.Email": TraitConstants(
            key: "exomind.base.Email",
            traitType: .email,
            name: nil,
            nameDefault: "Untitled email",
            icon: "envelope-o",
            color: 6,
            order: 4,
            collectionLike: false
    ),
    "exomind.base.Collection": TraitConstants(
            key: "exomind.base.Collection",
            traitType: .collection,
            name: nil,
            nameDefault: nil,
            icon: "folder-o",
            color: 2,
            order: 5,
            collectionLike: true
    ),
    "exomind.base.Task": TraitConstants(
            key: "exomind.base.Task",
            traitType: .task,
            name: nil,
            nameDefault: nil,
            icon: "check-square-o",
            color: 7,
            order: 6,
            collectionLike: false
    ),
    "exomind.base.Note": TraitConstants(
            key: "exomind.base.Note",
            traitType: .note,
            name: nil,
            nameDefault: nil,
            icon: "pencil",
            color: 3,
            order: 7,
            collectionLike: false
    ),
    "exomind.base.Link": TraitConstants(
            key: "exomind.base.Link",
            traitType: .link,
            name: nil,
            nameDefault: "Untitled link",
            icon: "link",
            color: 9,
            order: 8,
            collectionLike: false
    )
]